### DISCUSSION TITLE
Do I need to link ops on the same FD?

### DISCUSSION THREAD

**Discussion by redbaron** [axboe/liburing#843] [Q&A] [answered]
Are operations on same file descriptor implicitly serialized or not? For example, if I write and then read, should these ops be linked?

I know there can be a partial write,so it is probably not a good practice in real life, but I am just trying to understand what io_uring offers.

**Comment by axboe** [comment, marked as answer]
Nothing is serialized unless you ask for it to be, doesn't matter if the ops touch the same or different files.

Partial writes will be considered an error in terms of ```IOSQE_IO_LINK```, so it would break the chain.

**Comment by redbaron** [comment]
I see. If fd is a TCP socket, partial write is expected behaviour, right? does it still break the chain?

**Comment by axboe** [comment]
As I wrote: "Partial writes will be considered an error in terms of IOSQE_IO_LINK, so it would break the chain.". ```IOSQE_IO_HARDLINK``` does not sever on errors or short IO, but ```IOSQE_IO_LINK``` does.

**Comment by isilence** [comment]
Just as Jens said for write requests. That also holds for all kinds of send requests (as well as recv) when you set `MSG_WAITALL`, but otherwise partial IO is considered a "success" and doesn't break links.
